src/Elrond/TransactionFactory.cpp

```diff
@@ -0,0 +1,108 @@
+    this->networkConfig = networkConfig;
+}
+
+Proto::TransactionMessage TransactionFactory::createEGLDTransfer(const Address& sender, const Address& receiver, uint256_t amount) {
+    Proto::TransactionMessage message;
+
+    message.set_sender(sender.string());
+    message.set_receiver(receiver.string());
+    message.set_value(toString(amount));
+    message.set_gas_limit(this->networkConfig.getMinGasLimit());
+    message.set_gas_price(this->networkConfig.getMinGasPrice());
+    message.set_chain_id(this->networkConfig.getChainId());
+    message.set_version(1);
+
+    return message;
+}
+
+Proto::TransactionMessage TransactionFactory::createESDTTransfer(const Address& sender, const Address& receiver, const std::string& tokenIdentifier, uint256_t amount) {
+    Proto::TransactionMessage message;
+
+    std::string encodedTokenIdentifier = Codec::encodeStringTopLevel(tokenIdentifier);
+    std::string encodedAmount = Codec::encodeBigIntTopLevel(amount);
+    std::string data = prepareFunctionCall("ESDTTransfer", { encodedTokenIdentifier, encodedAmount });
+
+    // Additional gas to account for eventual increases in gas requirements (thus avoid breaking changes in clients of TW-core).
+    const long ADDITIONAL_GAS = 100000;
+
+    long gasLimit = 
+        this->networkConfig.getMinGasLimit() + 
+        this->networkConfig.getGasCostESDTTransfer() + 
+        this->networkConfig.getGasPerDataByte() * data.size() +
+        ADDITIONAL_GAS;
+
+    message.set_sender(sender.string());
+    message.set_receiver(receiver.string());
+    message.set_data(data);
+    message.set_gas_limit(gasLimit);
+    message.set_gas_price(this->networkConfig.getMinGasPrice());
+    message.set_chain_id(this->networkConfig.getChainId());
+    message.set_version(1);
+
+    return message;
+}
+
+Proto::TransactionMessage TransactionFactory::createESDTNFTTransfer(const Address& sender, const Address& receiver, const std::string& collection, uint64_t nonce, uint256_t quantity) {
+    Proto::TransactionMessage message;
+
+    std::string encodedCollection = Codec::encodeStringTopLevel(collection);
+    std::string encodedNonce = Codec::encodeUint64TopLevel(nonce);
+    std::string encodedQuantity = Codec::encodeBigIntTopLevel(quantity);
+    std::string encodedReceiver = Codec::encodeAddressTopLevel(receiver);
+    std::string data = prepareFunctionCall("ESDTNFTTransfer", { encodedCollection, encodedNonce, encodedQuantity, encodedReceiver });
+
+    // Additional gas to account for extra blockchain operations (e.g. data movement (between accounts) for NFTs), 
+    // and for eventual increases in gas requirements (thus avoid breaking changes in clients of TW-core).
+    const long ADDITIONAL_GAS = 500000;
+
+    long gasLimit = 
+        this->networkConfig.getMinGasLimit() + 
+        this->networkConfig.getGasCostESDTNFTTransfer() + 
+        this->networkConfig.getGasPerDataByte() * data.size() +
+        ADDITIONAL_GAS;
```

Comment:
Consider creating a new method on `networkConfig` named `computeGasLimitForESDTNFTTransfer()`, which receives `data.size()` as argument. This way, gas limit computations will be kept all in a single place.

Reply:
Defined a new class: `GasEstimator` (depends on `NetworkConfig`, used by `TransactionFactory`).